File: leetcode/solutions/0198.HouseRobber/rob.ts
export default function rob(nums: number[]): number {
  let dp = [nums[0], nums[1]];

  for (let i = 2; i < nums.length; i++) {
    let tmp = dp[1];
    dp[1] = Math.max(dp[0] + nums[i], dp[1]);
    dp[0] = tmp;
  }

  return Math.max(dp[0], dp[1]);
}

File: leetcode/solutions/0198.HouseRobber/rob.go
package solution

func rob(nums []int) int {
	dp := [2]int{}

	for _, num := range nums {
		dp[0], dp[1] = dp[1], max(dp[1], dp[0]+num)
	}

	return max(dp[0], dp[1])
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}
